File: netlify/functions/routes/preferences.ts
import { Hono } from "hono";
import { db } from "../lib/db.js";
import { requireUser, type PublicUser } from "../lib/auth.js";

type Variables = { user: PublicUser };
export const preferencesRoutes = new Hono<{ Variables: Variables }>();

const DEFAULT_PREFS = {
  telegram_chat_id: "",
  email_alerts: true,
  telegram_alerts: true,
  delivery_time: "07:00",
  language: "en",
  preferred_sectors: [] as string[],
  horizon: "both",
  risk_appetite: "medium",
  watchlist: [] as string[],
};

const ALLOWED_KEYS = new Set(Object.keys(DEFAULT_PREFS));

preferencesRoutes.get("/", requireUser, async (c) => {
  const user = c.get("user");
  const { data } = await db.from("user_preferences").select("*").eq("user_id", user.id).maybeSingle();
  if (!data) {
    return c.json({ ...DEFAULT_PREFS, user_id: user.id });
  }
  return c.json(data);
});

preferencesRoutes.put("/", requireUser, async (c) => {
  const user = c.get("user");
  const body = await c.req.json<Record<string, unknown>>();
  const patch: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(body)) {
    if (ALLOWED_KEYS.has(k)) patch[k] = v;
  }
  patch.updated_at = new Date().toISOString();

  const { data, error } = await db
    .from("user_preferences")
    .upsert({ user_id: user.id, ...patch }, { onConflict: "user_id" })
    .select("*")
    .single();

  if (error) return c.json({ detail: "Failed to update preferences" }, 500);
  return c.json(data);
});

File: legacy/python-fastapi/backend/routes/preferences.py
"""User preferences (per-user)."""
from datetime import datetime, timezone
from typing import Any, Dict

from fastapi import APIRouter, Depends

from auth import get_current_user
from db import preferences_col

router = APIRouter(prefix="/preferences", tags=["preferences"])


DEFAULT_PREFS = {
    "telegram_chat_id": "",
    "email_alerts": True,
    "telegram_alerts": True,
    "delivery_time": "07:00",
    "language": "en",
    "preferred_sectors": [],
    "horizon": "both",
    "risk_appetite": "medium",
    "watchlist": [],
}


@router.get("")
async def get_prefs(user: dict = Depends(get_current_user)) -> Dict[str, Any]:
    doc = await preferences_col.find_one({"user_id": user["id"]}, {"_id": 0})
    if not doc:
        doc = {**DEFAULT_PREFS, "user_id": user["id"]}
    return doc


@router.put("")
async def update_prefs(body: Dict[str, Any], user: dict = Depends(get_current_user)) -> Dict[str, Any]:
    allowed = set(DEFAULT_PREFS.keys())
    patch = {k: v for k, v in body.items() if k in allowed}
    patch["updated_at"] = datetime.now(timezone.utc).isoformat()
    await preferences_col.update_one(
        {"user_id": user["id"]},
        {"$set": {"user_id": user["id"], **patch}},
        upsert=True,
    )
    doc = await preferences_col.find_one({"user_id": user["id"]}, {"_id": 0})
    return doc
